baseplate/server/runtime_monitor.py

```diff
@@ -32,12 +37,36 @@ def report(self, batch: metrics.Batch) -> None:
         raise NotImplementedError
 
 
-class _ConcurrencyReporter(_Reporter):
+class _OpenConnectionsReporter(_Reporter):
     def __init__(self, pool: Pool):
         self.pool = pool
 
     def report(self, batch: metrics.Batch) -> None:
-        batch.gauge("active_requests").replace(len(self.pool.greenlets))
+        batch.gauge("open_connections").replace(len(self.pool.greenlets))
+
+
+class _ActiveRequestsObserver(BaseplateObserver, _Reporter):
+    def __init__(self):
+        self.live_requests = set()
+
+    def on_server_span_created(self, context: "RequestContext", server_span: "ServerSpan") -> None:
+        observer = _ActiveRequestsServerSpanObserver(self, server_span.trace_id)
+        server_span.register(observer)
+
+    def report(self, batch: metrics.Batch) -> None:
+        batch.gauge("active_requests").replace(len(self.live_requests))
+
+
+class _ActiveRequestsServerSpanObserver(ServerSpanObserver):
+    def __init__(self, reporter: _ActiveRequestsObserver, trace_id: int):
+        self.reporter = reporter
+        self.trace_id = trace_id
+
+    def on_start(self) -> None:
+        self.reporter.live_requests.add(self.trace_id)
```

Comment:
There's a non-zero (but hopefully very close to zero) chance of this leaving requests counting even though they are long gone. I was considering storing a timestamp and periodically sweeping out very old ones or something similar, but felt that's probably overcomplicated since processes restart occasionally anyway. Thoughts?

Reply:
> processes restart occasionally anyway

That's true in the legacy deployments, but not in k8s.  In k8s, they only restart if they crash or when there's a deploy (although deploy's aren't technically a restart in k8s like they are in legacy deployments).

Reply:
indeed, I still think that should change.

Reply:
> I was considering storing a timestamp and periodically sweeping out very old ones or something similar

How about `self.live_requests: Dict[int, Set[int]] = defaultdict(set)` and key-ing that using `int(time.time()`?  Then you can just `pop` any keys less than or equal to `int(time.time()) - 15` (or however many seconds you want to keep around) at the top of `report` and use something like `len(list(itertools.chain(self.live_requests.values()))` to get the count?

Reply:
```python
class _ActiveRequestsObserver(BaseplateObserver, _Reporter):
    def __init__(self) -> None:
        self.live_requests: Dict[int, Set[int]] = defaultdict(set)

    def add(self, trace_id: int) -> int:
        key = int(time.time())
        self.live_requests[key].add(trace_id)
        return key

    def remove(self, timestamp: int, trace_id: int) -> None:
        if timestamp in self.live_requests:
            self.live_requests[timestamp].remove(trace_id)

    def on_server_span_created(self, context: RequestContext, server_span: ServerSpan) -> None:
        observer = _ActiveRequestsServerSpanObserver(self, server_span.trace_id)
        server_span.register(observer)

    def report(self, batch: metrics.Batch) -> None:
        cutoff = int(time.time()) - 15
        for expired in filter(lambda k: k <= cutoff, self.live_requests.keys()):
            self.live_requests.pop(expired, None)
        batch.gauge("active_requests").replace(len(list(itertools.chain(self.live_requests.values()))))
```

Reply:
Yeah I was just doing `self.live_requests[trace_id] = time.time()` and then scanning them away but `time.time()` isn't free so I felt iffy on the value.

Reply:
That `15` is also a little sketch to me especially with request timeouts coming in at the same time. Ideally we'd want to cull things only after their request timeout, but that gets real complex.